pkg/cfn/builder/vpc_existing.go

```diff
@@ -0,0 +1,167 @@
+package builder
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	awsec2 "github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/pkg/errors"
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	gfnt "github.com/weaveworks/goformation/v4/cloudformation/types"
+)
+
+type ExistingVPCResourceSet struct {
+	rs            *resourceSet
+	clusterConfig *api.ClusterConfig
+	ec2API        ec2iface.EC2API
+	vpcID         *gfnt.Value
+	subnetDetails *SubnetDetails
+}
+
+// NewExistingVPCResourceSet creates and returns a new VPCResourceSet
+func NewExistingVPCResourceSet(rs *resourceSet, clusterConfig *api.ClusterConfig, ec2API ec2iface.EC2API) *ExistingVPCResourceSet {
+	return &ExistingVPCResourceSet{
+		rs:            rs,
+		clusterConfig: clusterConfig,
+		ec2API:        ec2API,
+		vpcID:         gfnt.NewString(clusterConfig.VPC.ID),
+		subnetDetails: &SubnetDetails{},
+	}
+}
+
+func (v *ExistingVPCResourceSet) CreateTemplate() (*gfnt.Value, *SubnetDetails, error) {
+	if v.clusterConfig.VPC.IPFamily == api.IPV6Family {
+		if err := v.checkIPv6CidrBlockAssociated(); err != nil {
+			return nil, nil, err
+		}
+	}
+	if err := v.importExistingResources(); err != nil {
+		return nil, nil, errors.Wrap(err, "error importing VPC resources")
+	}
+	return v.vpcID, v.subnetDetails, nil
+}
+
+func (v *ExistingVPCResourceSet) checkIPv6CidrBlockAssociated() error {
+	out, err := v.ec2API.DescribeVpcs(&awsec2.DescribeVpcsInput{
+		VpcIds: aws.StringSlice([]string{v.clusterConfig.VPC.ID}),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to describe VPC %q: %w", v.clusterConfig.VPC.ID, err)
+	}
+	if len(out.Vpcs) == 0 {
+		return fmt.Errorf("VPC %q does not exist", v.clusterConfig.VPC.ID)
+	}
+	if len(out.Vpcs[0].Ipv6CidrBlockAssociationSet) == 0 {
+		return fmt.Errorf("VPC %q does not have any associated IPv6 cidr blocks", v.clusterConfig.VPC.ID)
+	}
+	return nil
+}
+
+func (v *ExistingVPCResourceSet) importExistingResources() error {
+	if subnets := v.clusterConfig.VPC.Subnets.Private; subnets != nil {
+		var (
+			subnetRoutes map[string]string
+			err          error
+		)
+		if v.isFullyPrivate() {
+			subnetRoutes, err = importRouteTables(v.ec2API, v.clusterConfig.VPC.Subnets.Private)
+			if err != nil {
+				return err
+			}
+		}
+
+		subnetResources, err := makeSubnetResources(subnets, subnetRoutes)
+		if err != nil {
+			return err
+		}
+		v.subnetDetails.Private = subnetResources
+	}
+
+	if subnets := v.clusterConfig.VPC.Subnets.Public; subnets != nil {
+		subnetResources, err := makeSubnetResources(subnets, nil)
+		if err != nil {
+			return err
+		}
+		v.subnetDetails.Public = subnetResources
+	}
+
+	return nil
+}
+
+func makeSubnetResources(subnets map[string]api.AZSubnetSpec, subnetRoutes map[string]string) ([]SubnetResource, error) {
+	subnetResources := make([]SubnetResource, len(subnets))
+	i := 0
```

Comment:
Is there a reason why you aren't using `for i, network := range subnets {` instead of declaring this `i` here?

Reply:
copy and pasted the existing into a new file :smile: I'll refactor though :smile:

Reply:
oh its a map, it had no index :ahhhh: :smile:

Reply:
Ahhhh what the fuck :DDDDD